integration/querydsl/blaze-persistence-querydsl-expressions/src/main/java/com/blazebit/persistence/querydsl/BlazeCriteriaBuilderRenderer.java

```diff
@@ -0,0 +1,1308 @@
+                }
+            });
+        }
+    }
+
+    private void renderDistinct(QueryMetadata metadata, DistinctBuilder<?> criteriaBuilder) {
+        if (metadata.isDistinct()) {
+            criteriaBuilder.distinct();
+        }
+    }
+
+    private <X extends FromBuilder<X>> X renderJoins(QueryMetadata metadata, FromBaseBuilder<X> fromBuilder) {
+        X criteriaBuilder = null;
+        for (final JoinExpression joinExpression : metadata.getJoins()) {
+            boolean fetch = joinExpression.hasFlag(JPAQueryMixin.FETCH);
+            boolean hasCondition = joinExpression.getCondition() != null;
+            Expression<?> target = joinExpression.getTarget();
+            String alias = null;
+
+            if (target instanceof Operation<?>) {
+                Operation<?> operation = (Operation<?>) target;
+                if (operation.getOperator() == Ops.ALIAS) {
+                    target = operation.getArg(0);
+                    alias = ((Path<?>) operation.getArg(1)).getMetadata().getName();
+                }
+            }
+
+            if (target instanceof ValuesExpression<?>) {
+                ValuesExpression<?> valuesExpression = (ValuesExpression<?>) target;
+                Class type = valuesExpression.getRoot().getType();
+                String name = valuesExpression.getAlias().getMetadata().getName();
+                Collection<?> elements = valuesExpression.getElements();
+
+                if (! valuesExpression.getMetadata().isRoot()) {
+                    String attribute = relativePathString(valuesExpression.getRoot(), valuesExpression);
+                    if ( valuesExpression.isIdentifiable() ) {
+                        criteriaBuilder = (X) fromBuilder.fromIdentifiableValues(type, attribute, name, elements);
+                    } else {
+                        criteriaBuilder = (X) fromBuilder.fromValues(type, attribute, name, elements);
+                    }
+                } else if ( valuesExpression.isIdentifiable() ) {
+                    criteriaBuilder = (X) fromBuilder.fromIdentifiableValues(type, name, elements);
+                } else {
+                    criteriaBuilder = (X) fromBuilder.fromValues(type, name, elements);
+                }
+            } else if (target instanceof Path<?>) {
+                Path<?> entityPath = (Path<?>) target;
+                if (alias == null) {
+                    alias = entityPath.getMetadata().getName();
+                }
+                boolean entityJoin = entityPath.getMetadata().isRoot();
+
+                switch (joinExpression.getType()) {
+                    case DEFAULT:
+                        if (fromBuilder instanceof FromBuilder) {
+                            criteriaBuilder = (X) fromBuilder;
+                            From from = criteriaBuilder.getFrom(alias);
+
+                            // TODO find a clean way to detect pre-set FROM clauses...
```

Comment:
The issue we talked about, that I currently need to instantiate a CBF with return type `Object.class` because I need to reliably remove/detect the default join target.

Reply:
Hmm, if you always invoke a from method, there will be no default from element, so that might be good enough to clear this one.

Reply:
Actually, this is to detect the from clause set by `fromEntitySubquery`

Reply:
Not sure I understand, I didn't dig that much into it yet. Just wanted to make sure you give this another thought but I am sure you know what you are doing ;)